File: AliyunOSSiOSTests/OSSConfigurationTest.m
//
//  OSSConfigurationTest.m
//  AliyunOSSiOSTests
//
//  Created by ws on 2021/3/17.
//  Copyright © 2021 aliyun. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <AliyunOSSiOS/AliyunOSSiOS.h>
#import <AliyunOSSiOS/OSSServiceSignature.h>
#import <AliyunOSSiOS/NSData+OSS.h>
#import "OSSTestMacros.h"
#import "OSSTestUtils.h"

@interface OSSConfigurationTest : XCTestCase {
    NSString *host;
    NSString *scheme;
    NSString *endpoint;
    NSString *cname;
    NSString *cnameEndpoint;
    NSString *bucketEndpoint;
}

@end

@implementation OSSConfigurationTest

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.
    host = @"oss-cn-beijing.aliyuncs.com";
    cname = @"oss.cname.com";
    scheme = @"https://";
    endpoint = [NSString stringWithFormat:@"%@%@", scheme, host];
    cnameEndpoint = [NSString stringWithFormat:@"%@%@", scheme, cname];
    bucketEndpoint = [NSString stringWithFormat:@"%@.%@", OSS_BUCKET_PUBLIC, host];
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testDefault {
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:endpoint credentialProvider:credentialProvider clientConfiguration:config];
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    get.objectKey = OSS_MULTIPART_UPLOADKEY;
    [[[client getObject:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertTrue([task.error.userInfo[@"HostId"] isEqualToString:bucketEndpoint]);
        XCTAssertTrue([task.error.userInfo[@"Bucket"] isEqualToString:OSS_BUCKET_PUBLIC]);
        return task;
    }] waitUntilFinished];
}

- (void)testPathStyleAccessEnable {
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.maxRetryCount = 0;
    config.isPathStyleAccessEnable = YES;
    config.cnameExcludeList = @[cname];
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:cnameEndpoint credentialProvider:credentialProvider clientConfiguration:config];
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    get.objectKey = OSS_MULTIPART_UPLOADKEY;
    [[[client getObject:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertNotNil(task.error);
        NSString *url = [NSString stringWithFormat:@"%@%@/%@/%@", scheme, cname, OSS_BUCKET_PUBLIC, OSS_MULTIPART_UPLOADKEY];
        XCTAssertTrue([task.error.userInfo[@"NSErrorFailingURLStringKey"] isEqualToString:url]);
        return task;
    }] waitUntilFinished];
}

- (void)testSupportCnameEnable {
    NSArray *cnameExcludeList = @[cname];
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.maxRetryCount = 0;
    config.cnameExcludeList = cnameExcludeList;
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:cnameEndpoint credentialProvider:credentialProvider clientConfiguration:config];
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    get.objectKey = OSS_MULTIPART_UPLOADKEY;
    [[[client getObject:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertNotNil(task.error);
        NSString *url = [NSString stringWithFormat:@"%@%@.%@/%@", scheme, OSS_BUCKET_PUBLIC, cname, OSS_MULTIPART_UPLOADKEY];
        XCTAssertTrue([task.error.userInfo[@"NSErrorFailingURLStringKey"] isEqualToString:url]);
        return task;
    }] waitUntilFinished];
    
    config = [OSSClientConfiguration new];
    config.maxRetryCount = 0;
    client = [[OSSClient alloc] initWithEndpoint:cnameEndpoint credentialProvider:credentialProvider clientConfiguration:config];
    get = [OSSGetObjectRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    get.objectKey = OSS_MULTIPART_UPLOADKEY;
    [[[client getObject:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertNotNil(task.error);
        NSString *url = [NSString stringWithFormat:@"%@%@/%@", scheme, cname, OSS_MULTIPART_UPLOADKEY];
        XCTAssertTrue([task.error.userInfo[@"NSErrorFailingURLStringKey"] isEqualToString:url]);
        return task;
    }] waitUntilFinished];
}

- (void)testCustomPathPrefixEnable {
    NSString *endpointPath = [NSString stringWithFormat:@"%@/%@", cnameEndpoint, OSS_BUCKET_PUBLIC];
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.maxRetryCount = 0;
    config.isCustomPathPrefixEnable = YES;
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:endpointPath credentialProvider:credentialProvider clientConfiguration:config];
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    get.objectKey = OSS_MULTIPART_UPLOADKEY;
    [[[client getObject:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertNotNil(task.error);
        NSString *url = [NSString stringWithFormat:@"%@/%@", endpointPath, OSS_MULTIPART_UPLOADKEY];
        XCTAssertTrue([task.error.userInfo[@"NSErrorFailingURLStringKey"] isEqualToString:url]);
        return task;
    }] waitUntilFinished];
}

- (void)testCustomPathPrefixEnableWithNoPathEndpont {
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.maxRetryCount = 0;
    config.isCustomPathPrefixEnable = YES;
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:cnameEndpoint credentialProvider:credentialProvider clientConfiguration:config];
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    get.objectKey = OSS_MULTIPART_UPLOADKEY;
    [[[client getObject:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertNotNil(task.error);
        NSString *url = [NSString stringWithFormat:@"%@/%@", cnameEndpoint, OSS_MULTIPART_UPLOADKEY];
        XCTAssertTrue([task.error.userInfo[@"NSErrorFailingURLStringKey"] isEqualToString:url]);
        return task;
    }] waitUntilFinished];
}

- (void)testCustomPathPrefixEnableWithNullObject {
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.maxRetryCount = 0;
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:cnameEndpoint credentialProvider:credentialProvider clientConfiguration:config];
    OSSGetBucketRequest *get = [OSSGetBucketRequest new];
    get.bucketName = OSS_BUCKET_PUBLIC;
    [[[client getBucket:get] continueWithBlock:^id _Nullable(OSSTask * _Nonnull task) {
        XCTAssertNotNil(task.error);
        XCTAssertTrue([task.error.userInfo[@"NSErrorFailingURLStringKey"] isEqualToString:[cnameEndpoint stringByAppendingString:@"/"]]);
        return task;
    }] waitUntilFinished];
}

- (void)testAllowNetworkMetricInfo {
    OSSClientConfiguration *config = [OSSClientConfiguration new];
    OSSAuthCredentialProvider *credentialProvider = [[OSSAuthCredentialProvider alloc] initWithAuthServerUrl:OSS_STSTOKEN_URL];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:OSS_ENDPOINT credentialProvider:credentialProvider clientConfiguration:config];
    
    NSString *privateBucketName = [OSSTestUtils getBucketName];
    OSSCreateBucketRequest *createBucket = [OSSCreateBucketRequest new];
    createBucket.bucketName = privateBucketName;
    OSSTask *task = [client createBucket:createBucket];
    [task waitUntilFinished];
    XCTAssertNil(((OSSResult *)task.result).metrics);

    OSSPutObjectRequest * put = [OSSPutObjectRequest new];
    put.bucketName = privateBucketName;
    put.objectKey = OSS_IMAGE_KEY;
    put.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"hasky" withExtension:@"jpeg"];
    task = [client putObject:put];
    [task waitUntilFinished];
    XCTAssertNil(((OSSResult *)task.result).metrics);
    
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = privateBucketName;
    get.objectKey = @"error";
    get.onRecieveData = ^(NSData * _Nonnull data) {
    };
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error.userInfo[OSSNetworkTaskMetrics]);
    
    config = [OSSClientConfiguration new];
    config.isAllowNetworkMetricInfo = YES;
    client = [[OSSClient alloc] initWithEndpoint:OSS_ENDPOINT credentialProvider:credentialProvider clientConfiguration:config];
    
    put = [OSSPutObjectRequest new];
    put.bucketName = privateBucketName;
    put.objectKey = OSS_IMAGE_KEY;
    put.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"hasky" withExtension:@"jpeg"];
    task = [client putObject:put];
    [task waitUntilFinished];
    XCTAssertNotNil(((OSSResult *)task.result).metrics);
    
    get = [OSSGetObjectRequest new];
    get.bucketName = privateBucketName;
    get.objectKey = @"error";
    get.onRecieveData = ^(NSData * _Nonnull data) {
    };
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNotNil(task.error.userInfo[OSSNetworkTaskMetrics]);
    
    [OSSTestUtils cleanBucket:privateBucketName with:client];
}

- (void)testAPI_signerV1 {
    NSString *bucketName = [@"test-signerv1-" stringByAppendingFormat:@"%@", @((NSInteger)[NSDate new].timeIntervalSince1970)];
    NSString *objectKey = @"signerV1";
    NSURL * fileURL = [[NSBundle mainBundle] URLForResource:@"hasky" withExtension:@"jpeg"];

    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.signVersion = OSSSignVersionV1;
    
    // OSSPlainTextAKSKPairCredentialProvider
    id<OSSCredentialProvider> credentialProvider = [[OSSPlainTextAKSKPairCredentialProvider alloc] initWithPlainTextAccessKey:OSS_ACCESSKEY_ID
                                                                                                                    secretKey:OSS_SECRETKEY_ID];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:endpoint
                                         credentialProvider:credentialProvider
                                        clientConfiguration:config];
    
    OSSCreateBucketRequest *createBucketRequest = [OSSCreateBucketRequest new];
    createBucketRequest.bucketName = bucketName;
    OSSTask *task = [client createBucket:createBucketRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    OSSPutObjectRequest *putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    // OSSStsTokenCredentialProvider
    OSSFederationToken *federationToken = [OSSTestUtils getOssFederationToken];
    credentialProvider = [[OSSStsTokenCredentialProvider alloc] initWithAccessKeyId:federationToken.tAccessKey
                                                                        secretKeyId:federationToken.tSecretKey
                                                                      securityToken:federationToken.tToken];
    client = [[OSSClient alloc] initWithEndpoint:endpoint
                              credentialProvider:credentialProvider
                             clientConfiguration:config];
    
    putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    
    // OSSFederationCredentialProvider
    credentialProvider = [[OSSFederationCredentialProvider alloc] initWithFederationTokenGetter:^OSSFederationToken * _Nullable{
        return [OSSTestUtils getOssFederationToken];
    }];
    client = [[OSSClient alloc] initWithEndpoint:endpoint
                              credentialProvider:credentialProvider
                             clientConfiguration:config];
    
    putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    
    // OSSCustomSignerCredentialProvider
    credentialProvider = [[OSSCustomSignerCredentialProvider alloc] initWithImplementedSigner:^NSString * _Nullable(NSString * _Nonnull contentToSign, NSError *__autoreleasing  _Nullable * _Nullable error) {
        NSString *signature = [OSSUtil calBase64Sha1WithData:contentToSign withSecret:OSS_SECRETKEY_ID];
        XCTAssertNotNil(signature);
        return [NSString stringWithFormat:@"OSS %@:%@", OSS_ACCESSKEY_ID, signature];
    }];
    client = [[OSSClient alloc] initWithEndpoint:endpoint
                              credentialProvider:credentialProvider
                             clientConfiguration:config];
    
    putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    [OSSTestUtils cleanBucket:bucketName with:client];
}

- (void)testAPI_signerV4 {
    NSString *bucketName = [@"test-signerv4-" stringByAppendingFormat:@"%@", @((NSInteger)[NSDate new].timeIntervalSince1970)];
    NSString *objectKey = @"signerV4/!@#$%^&*()_=\\|';:><[.-+]{}?\"~`";
    NSURL * fileURL = [[NSBundle mainBundle] URLForResource:@"hasky" withExtension:@"jpeg"];
    NSString *endpoint = OSS_ENDPOINT;

    OSSClientConfiguration *config = [OSSClientConfiguration new];
    config.signVersion = OSSSignVersionV4;
    
    // OSSPlainTextAKSKPairCredentialProvider
    id<OSSCredentialProvider> credentialProvider = [[OSSPlainTextAKSKPairCredentialProvider alloc] initWithPlainTextAccessKey:OSS_ACCESSKEY_ID
                                                                                                                    secretKey:OSS_SECRETKEY_ID];
    OSSClient *client = [[OSSClient alloc] initWithEndpoint:endpoint
                                         credentialProvider:credentialProvider
                                        clientConfiguration:config];
    client.region = OSS_REGION;
    
    OSSCreateBucketRequest *createBucketRequest = [OSSCreateBucketRequest new];
    createBucketRequest.bucketName = bucketName;
    OSSTask *task = [client createBucket:createBucketRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    OSSPutObjectRequest *putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    OSSGetObjectRequest *get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    // OSSStsTokenCredentialProvider
    OSSFederationToken *federationToken = [OSSTestUtils getOssFederationToken];
    credentialProvider = [[OSSStsTokenCredentialProvider alloc] initWithAccessKeyId:federationToken.tAccessKey
                                                                        secretKeyId:federationToken.tSecretKey
                                                                      securityToken:federationToken.tToken];
    client = [[OSSClient alloc] initWithEndpoint:endpoint
                              credentialProvider:credentialProvider
                             clientConfiguration:config];
    client.region = OSS_REGION;
    
    putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    
    // OSSFederationCredentialProvider
    credentialProvider = [[OSSFederationCredentialProvider alloc] initWithFederationTokenGetter:^OSSFederationToken * _Nullable{
        return [OSSTestUtils getOssFederationToken];
    }];
    client = [[OSSClient alloc] initWithEndpoint:endpoint
                              credentialProvider:credentialProvider
                             clientConfiguration:config];
    client.region = OSS_REGION;

    putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNil(task.error);
    
    
    // OSSCustomSignerCredentialProvider
    credentialProvider = [[OSSCustomSignerCredentialProvider alloc] initWithImplementedSigner:^NSString * _Nullable(NSString * _Nonnull contentToSign, NSError *__autoreleasing  _Nullable * _Nullable error) {
        NSData *jsonData = [contentToSign dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *content = [NSJSONSerialization JSONObjectWithData:jsonData
                                                                options:NSJSONReadingMutableContainers
                                                                  error:error];
        NSString *date = content[OSSContentDate];
        NSString *region = content[OSSContentRegion];
        NSString *product = content[OSSContentProduct];
        NSString *stringToSign = content[OSSContentStringToSign];
        id<OSSServiceSignature> serviceSignature = [HmacSHA256Signature new];
        NSData *signingSecret = [[@"aliyun_v4" stringByAppendingString:OSS_SECRETKEY_ID] dataUsingEncoding:NSUTF8StringEncoding];
        NSData *signingDate = [serviceSignature computeHash:signingSecret
                                                       data:[date dataUsingEncoding:NSUTF8StringEncoding]];
        NSData *signingRegion = [serviceSignature computeHash:signingDate
                                                         data:[region dataUsingEncoding:NSUTF8StringEncoding]];
        NSData *signingService = [serviceSignature computeHash:signingRegion
                                                          data:[product dataUsingEncoding:NSUTF8StringEncoding]];
        NSData *signingKey =[serviceSignature computeHash:signingService
                                                     data:[@"aliyun_v4_request" dataUsingEncoding:NSUTF8StringEncoding]];
        NSData *result = [serviceSignature computeHash:signingKey
                                                  data:[stringToSign dataUsingEncoding:NSUTF8StringEncoding]];
        NSString *signature = [result oss_hexString];
        NSString *credential = [NSString stringWithFormat:@"Credential=%@/%@/%@/%@/aliyun_v4_request", OSS_ACCESSKEY_ID, date, region, product];
        NSString *signedHeaders = @"";
        NSString *sign = [NSString stringWithFormat:@",Signature=%@", signature];
        return [NSString stringWithFormat:@"OSS4-HMAC-SHA256 %@%@%@", credential, signedHeaders, sign];
    }];
    client = [[OSSClient alloc] initWithEndpoint:endpoint
                              credentialProvider:credentialProvider
                             clientConfiguration:config];
    client.region = OSS_REGION;

    putObjectRequest = [OSSPutObjectRequest new];
    putObjectRequest.bucketName = bucketName;
    putObjectRequest.objectKey = objectKey;
    putObjectRequest.uploadingFileURL = fileURL;
    task = [client putObject:putObjectRequest];
    [task waitUntilFinished];
    XCTAssertNotNil(task.error);
    XCTAssertTrue([task.error.userInfo[OSSErrorMessageTOKEN] isEqualToString:@"V4 signature does not support OSSCustomSignerCredentialProvider"]);

    get = [OSSGetObjectRequest new];
    get.bucketName = bucketName;
    get.objectKey = objectKey;
    task = [client getObject:get];
    [task waitUntilFinished];
    XCTAssertNotNil(task.error);
    XCTAssertTrue([task.error.userInfo[OSSErrorMessageTOKEN] isEqualToString:@"V4 signature does not support OSSCustomSignerCredentialProvider"]);

    [OSSTestUtils cleanBucket:bucketName with:client];
}

@end

File: OSSSwiftDemo/OSSSwiftDemoTests/OSSConfigurationTest.swift
//
//  OSSConfigurationTest.swift
//  OSSSwiftDemoTests
//
//  Created by ws on 2021/3/29.
//  Copyright © 2021 aliyun. All rights reserved.
//

import XCTest
import AliyunOSSiOS

class OSSConfigurationTest: XCTestCase {

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testExample() throws {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }

    func testPerformanceExample() throws {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
    func testDefault() {
        let config = OSSClientConfiguration()
        let credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        let client = OSSClient(endpoint: ENDPOINT, credentialProvider: credentialProvider, clientConfiguration: config)
        let get = OSSGetObjectRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        get.objectKey = OSS_MULTIPART_UPLOADKEY
        client.getObject(get).continue({ (task) -> Any? in
            if let user = task.error?._userInfo as? [String: String] {
                XCTAssertTrue(user["HostId"] == "\(OSS_BUCKET_PUBLIC).\(ENDPOINT)")
                XCTAssertTrue(user["BucketName"] == OSS_BUCKET_PUBLIC)
            } else {
                XCTAssertTrue(false)
            }
            
            return task
        }).waitUntilFinished()
    }

    func testPathStyleAccessEnable() {
        let config = OSSClientConfiguration()
        config.maxRetryCount = 0
        config.isPathStyleAccessEnable = true
        config.cnameExcludeList = [CNAME_ENDPOINT]
        let credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        let client = OSSClient(endpoint: "https://\(CNAME_ENDPOINT)", credentialProvider: credentialProvider, clientConfiguration: config)
        let get = OSSGetObjectRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        get.objectKey = OSS_MULTIPART_UPLOADKEY
        client.getObject(get).continue({ (task) -> Any? in
            XCTAssertNotNil(task.error);
            if let user = task.error?._userInfo as? [String: Any],
               let urlString = user["NSErrorFailingURLStringKey"] as? String {
                let url = "\(SCHEME)\(CNAME_ENDPOINT)/\(OSS_BUCKET_PUBLIC)/\(OSS_MULTIPART_UPLOADKEY)"
                XCTAssertTrue(urlString == url)
            } else {
                XCTAssertTrue(false)
            }
            return task
        }).waitUntilFinished()
    }
    
    func testSupportCnameEnable() {
        var config = OSSClientConfiguration()
        config.maxRetryCount = 0
        config.cnameExcludeList = [CNAME_ENDPOINT]
        var credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        var client = OSSClient(endpoint: "https://\(CNAME_ENDPOINT)", credentialProvider: credentialProvider, clientConfiguration: config)
        var get = OSSGetObjectRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        get.objectKey = OSS_MULTIPART_UPLOADKEY
        client.getObject(get).continue({ (task) -> Any? in
            XCTAssertNotNil(task.error);
            if let user = task.error?._userInfo as? [String: Any],
               let urlString = user["NSErrorFailingURLStringKey"] as? String {
                let url = "\(SCHEME)\(OSS_BUCKET_PUBLIC).\(CNAME_ENDPOINT)/\(OSS_MULTIPART_UPLOADKEY)"
                XCTAssertTrue(urlString == url)
            } else {
                XCTAssertTrue(false)
            }
            return task
        }).waitUntilFinished()
        
        config = OSSClientConfiguration()
        config.maxRetryCount = 0
        credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        client = OSSClient(endpoint: "https://\(CNAME_ENDPOINT)", credentialProvider: credentialProvider, clientConfiguration: config)
        get = OSSGetObjectRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        get.objectKey = OSS_MULTIPART_UPLOADKEY
        client.getObject(get).continue({ (task) -> Any? in
            XCTAssertNotNil(task.error);
            if let user = task.error?._userInfo as? [String: Any],
               let urlString = user["NSErrorFailingURLStringKey"] as? String {
                let url = "\(SCHEME)\(CNAME_ENDPOINT)/\(OSS_MULTIPART_UPLOADKEY)"
                XCTAssertTrue(urlString == url)
            } else {
                XCTAssertTrue(false)
            }
            return task
        }).waitUntilFinished()
    }
    
    func testCustomPathPrefixEnable() {
        let endpointPath = "https://\(CNAME_ENDPOINT)/path"
        let config = OSSClientConfiguration()
        config.maxRetryCount = 0
        config.isCustomPathPrefixEnable = true
        config.cnameExcludeList = [CNAME_ENDPOINT]
        let credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        let client = OSSClient(endpoint: endpointPath, credentialProvider: credentialProvider, clientConfiguration: config)
        let get = OSSGetObjectRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        get.objectKey = OSS_MULTIPART_UPLOADKEY
        client.getObject(get).continue({ (task) -> Any? in
            XCTAssertNotNil(task.error);
            if let user = task.error?._userInfo as? [String: Any],
               let urlString = user["NSErrorFailingURLStringKey"] as? String {
                let url = "\(endpointPath)/\(OSS_MULTIPART_UPLOADKEY)"
                XCTAssertTrue(urlString == url)
            } else {
                XCTAssertTrue(false)
            }
            return task
        }).waitUntilFinished()
    }
    
    func testCustomPathPrefixEnableWithNoPathEndpont() {
        let config = OSSClientConfiguration()
        config.maxRetryCount = 0
        config.isCustomPathPrefixEnable = true
        let credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        let client = OSSClient(endpoint: "https://\(CNAME_ENDPOINT)", credentialProvider: credentialProvider, clientConfiguration: config)
        let get = OSSGetObjectRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        get.objectKey = OSS_MULTIPART_UPLOADKEY
        client.getObject(get).continue({ (task) -> Any? in
            XCTAssertNotNil(task.error);
            if let user = task.error?._userInfo as? [String: Any],
               let urlString = user["NSErrorFailingURLStringKey"] as? String {
                let url = "\(SCHEME)\(CNAME_ENDPOINT)/\(OSS_MULTIPART_UPLOADKEY)"
                XCTAssertTrue(urlString == url)
            } else {
                XCTAssertTrue(false)
            }
            return task
        }).waitUntilFinished()
    }
    
    func testCustomPathPrefixEnableWithNullObject() {
        let config = OSSClientConfiguration()
        config.maxRetryCount = 0
        let credentialProvider = OSSAuthCredentialProvider(authServerUrl: OSS_STSTOKEN_URL)
        let client = OSSClient(endpoint: "https://\(CNAME_ENDPOINT)", credentialProvider: credentialProvider, clientConfiguration: config)
        let get = OSSGetBucketRequest()
        get.bucketName = OSS_BUCKET_PUBLIC
        client.getBucket(get).continue({ (task) -> Any? in
            XCTAssertNotNil(task.error);
            if let user = task.error?._userInfo as? [String: Any],
               let urlString = user["NSErrorFailingURLStringKey"] as? String {
                let url = "\(SCHEME)\(CNAME_ENDPOINT)/"
                XCTAssertTrue(urlString == url)
            } else {
                XCTAssertTrue(false)
            }
            return task
        }).waitUntilFinished()
    }
}
